lib/grpc/stream.ex

```diff
@@ -190,19 +203,125 @@ defmodule GRPC.Stream do
     end
 
-    dry_run? = Keyword.get(opts, :dry_run, false)
-
     flow
-    |> Flow.map(fn msg ->
-      if not dry_run? do
-        send_response(from, msg)
-      end
+    |> Flow.map(fn
+      {:ok, msg} ->
+        send_response(from, msg, opts)
+
+        flow
+
+      {:error, %GRPC.RPCError{} = reason} ->
+        send_response(from, reason, opts)
+        flow
 
-      flow
+      {:error, reason} ->
+        msg = GRPC.RPCError.exception(message: "[Error] #{inspect(reason)}")
+        send_response(from, msg, opts)
+        flow
+
+      msg ->
+        send_response(from, msg, opts)
+        flow
     end)
     |> Flow.run()
   end
 
+  @doc """
+  Intercepts and transforms **error tuples** or **unexpected exceptions** that occur
+  within a gRPC stream pipeline.
+
+  `map_error/3` allows graceful handling or recovery from errors produced by previous
+  operators (e.g. `map/2`, `flat_map/2`) or from validation logic applied to incoming data.
+
+  The provided `handler/1` function receives the error reason (or the exception struct)
+  and can either:
+
+    * Return a new error tuple — e.g. `{:error, new_reason}` — to re-emit a modified error.
+    * Return any other value to recover from the failure and continue the pipeline.
+
+  This makes it suitable for both **input validation** and **capturing unexpected runtime errors**
+  in stream transformations.
+
+  ## Parameters
+
+    - `stream` — The input stream or `Flow` pipeline.
+    - `func` — A function that takes an error reason or exception and returns either a new value or an error tuple.
+
+  ## Returns
+
+    - A new stream where all error tuples and raised exceptions are processed by `func/1`.
+
+  ## Examples
+
+      iex> GRPC.Stream.from([1, 2])
+      ...> |> GRPC.Stream.map(fn
+      ...>   2 -> raise "boom"
+      ...>   x -> x
+      ...> end)
+      ...> |> GRPC.Stream.map_error(fn
+      ...>   {:error, _reason} ->
+      ...>     GRPC.RPCError.exception(message: "Validation or runtime error")
+      ...>   msg -> msg
+      ...> end)
+
+  In this example:
+
+  * The call to `GRPC.Stream.map/2` raises an exception for value `2`.
+  * `map_error/3` catches the error and wraps it in a `GRPC.RPCError` struct with a custom message.
+  * The pipeline continues execution, transforming errors into structured responses.
+
+  ## Notes
+
+  - `map_error/3` is **lazy** and only executes when the stream is materialized
+    (via `GRPC.Stream.run/1` or `GRPC.Stream.run_with/3`).
+
+  - Use this operator to implement **robust error recovery**, **input validation**, or
+    to normalize exceptions from downstream Flow stages into well-defined gRPC errors.
+  """
+  defdelegate map_error(stream, func), to: Operators
+
+  @doc """
+  Applies a **side-effecting function** to each element of the stream **without altering** its values.
+
+  The `effect/2` function is useful for performing **imperative or external actions** 
```

Comment:
since we have map, maybe `each` would be a nice counterpart since we have Enum.each and Enum.map

Reply:
You said it would be nice to have an `each` function, right? That's what I understood from your comment.
So I said that `each` would definitely have to materialize the stream (you can't have a stream operator after `each`). Then I said that we already have the equivalent of `each` in the `run` and `run_with` functions.

Reply:
I just meant that effect could be named each instead

Reply:
because of the confusion I made I don't think it's a good name hahaha